### PR TITLE
Models

### PR DESCRIPTION
**Pull request by viktor76525**

### CODE REVIEW

File: empirical/GMM_models/McVerry_2006_Sa.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    C43 = [-0.33716, -0.31036, -0.49068, -0.46604, -0.31282, -0.07565, 0.17615, 0.34775, 0.72380, 0.89239, 0.77743, 0.77743, 0.60938]
+    C46 = [-0.03255, -0.0325, -0.03441, -0.03594, -0.03823, -0.03535, -0.03354, -0.03211, -0.02857, -0.025, -0.02008, -0.02008, -0.01587]
+    Sigma6 = [0.4871, 0.5099, 0.5297, 0.5401, 0.5599, 0.5456, 0.5556, 0.5658, 0.5611, 0.5573, 0.5419, 0.5419, 0.5809]
+    Sigslope = [-0.1011, -0.0259, -0.0703, -0.0292, 0.0172, -0.0566, -0.1064, -0.1123, -0.0836, -0.0620, 0.0385, 0.0385, 0.1403]
+    Tau = [0.2677, 0.2469, 0.3139, 0.3017, 0.2583, 0.1967, 0.1802, 0.1440, 0.1871, 0.2073, 0.2405, 0.2405, 0.2053]
+    # fmt: on
+
+    M = faultprop.Mw
+    R = siteprop.Rrup
+    # not sure about this because it is not documented
```
Comment (jasonmotha) [inline empirical/GMM_models/McVerry_2006_Sa.py]: PGA is equivalent to a pSA period of 0. PGV is normally hacked to be -1.

So, may have to swap them around. Or have a look in here to figure it out http://www.nzsee.org.nz/db/Bulletin/Archive/39(1)0001.pdf

File: empirical/GMM_models/bc_hydro_2016_subduction.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        deltaC1 = np.interp(
+            np.log(periods),
+            np.log([1e-10, 0.3, 0.5, 1.0, 2.0, 3.0, 1e10]),
+            [0.2, 0.2, 0.1, 0.0, -0.1, -0.2, -0.2],
+        )
+        deltaC1[0] = 0.2
+
+    Vs_star = min(siteprop.vs30, 1000)
+
+    # seems silly to loop over all periods
```
Comment (jasonmotha) [inline empirical/GMM_models/bc_hydro_2016_subduction.py]: Very silly. Can it at least do multiple periods at once?

We should probably chat to Brendon and "fix" this code.

File: empirical/util/classdef.py [hunk 3]
```diff
@@ -28,20 +31,24 @@ def __init__(self):
         self.z2p5 = None  # (km)
         self.siteclass = None
         self.orientation = "average"
+        self.rvol = (
```
[the remark this answers is not in the document]
Comment (jasonmotha) [reply empirical/util/classdef.py]: From what I've read it is the same.

File: empirical/GMM_models/McVerry_2006_Sa.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    if im == "PGA":
+        period = -1
+    elif im == "PGV":
+        period = 0
+    assert period <= (periods[-1] + 0.0001)
+
+    # interpolate between periods if neccesary
```
[the remark this answers is not in the document]
Comment (viktor76525) [reply empirical/GMM_models/McVerry_2006_Sa.py]: no

File: empirical/GMM_models/bc_hydro_2016_subduction.py (hunk 2 above)
Comment (viktor76525) [reply empirical/GMM_models/bc_hydro_2016_subduction.py]: no, yup

File: empirical/GMM_models/ASK_2014_nga.py [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+            if ry0 - Ry1 <= 0:
+                t5 = 1
+            elif ry0 - Ry1 < 5:
+                t5 = 1 - (ry0 - Ry1) / 5
+            else:
+                t5 = 0
+
+        return a13[ip] * t1 * t2 * t3 * t4 * t5
+
+    f4 = hanging_wall()
```
[the remark this answers is not in the document]
Comment (viktor76525) [reply empirical/GMM_models/ASK_2014_nga.py]: automatic now

File: empirical/GMM_models/CY_2014_nga.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+        if 6 < M < 6.9:
+            term10 *= gamma_JP_IT[ip]
+    elif region == 3:
+        term10 *= gamma_Wn[ip]
+
+    # Style of faulting
+    term2 = (c1_a[ip] + c1_c[ip] / (math.cosh(2 * max(M - 4.5, 0)))) * F_RV
+    term3 = (c1_b[ip] + c1_d[ip] / math.cosh(2 * max(M - 4.5, 0))) * F_NM
+
+    # Ztor
```
[the remark this answers is not in the document]
Comment (viktor76525) [reply empirical/GMM_models/CY_2014_nga.py]: EZtor is always used in term4

File: empirical/GMM_models/McVerry_2006_Sa.py (hunk 1 above)
Comment (viktor76525) [reply empirical/GMM_models/McVerry_2006_Sa.py]: fixed

File: empirical/GMM_models/ASK_2014_nga.py [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+    elif crjb < 15:
+        f11 = a14[ip] * (1 - (crjb - 5) / 10)
+
+    # Sa
+    lnSa = (
+        f1 + f6 + f_rv * f7 + f_nm * f8 + f_hw * f4 + f_as * f11 + f5 + f10 + regional
+    )
+    Sa = np.exp(lnSa)
+
+    # Standard deviation
```
[the remark this answers is not in the document]
Comment (viktor76525) [reply empirical/GMM_models/ASK_2014_nga.py]: done, but it is not really viable for CY_2014

**Comment by viktor76525** [comment]
It's handy having the matlab/octave scripts next to the python test scripts because they are so linked, almost identical copies of each other, if one changes, so should the other. makes it easier to re-run with octave, python in a 2 step process.

**Pull request merged**